### ISSUE TITLE
Feature request: Different minimization criteria for MinimumAreaBox2

### ISSUE THREAD

**Issue by dempo93** [davideberly/GeometricTools#37]
Hello,
First of all, thanks for the amazing work on this library!

I recently faced the problem of rotating and aligning sets of points so that they occupy the least space along a single dimension, in the image below the "rectangle by width" case. 

![minbox](https://user-images.githubusercontent.com/19284225/151814621-595c1175-4e08-41e0-84ac-6afa6165cbdb.jpg)
(image courtesy of ArcGIS)

Would it be possible for you to support such cases? For what I saw you already have an implementation of the rotating calipers algorithm, and the only relevant change would be to update the minimization criterium in MinimumAreaBox2.h from

```
template <typename InputType, typename ComputeType>
class MinimumAreaBox2 {
....
    if (box.area < minBox.area)
    {
         minBox = box;
    }
...
};
```
to something like

```
template <typename InputType, typename ComputeType, typename BoxComparatorType>
class MinimumBox2 {
....
    BoxComparatorType comparator{};
    if (comparator(box.area,minBox.area))
    {
         minBox = box;
    }
...
};
```
and then provide for example an `AreaLessComparator` and a `SideLessComparator`.

**Comment by davideberly** [comment]
Yes, I will add that feature to GTE. I have made a note of the request at my website in the [Technical Support Queue document](https://www.geometrictools.com/Downloads/TechnicalSupportQueue.pdf).

**Comment by davideberly** [comment]
This support item is now the active one in my queue. The approach of adding BoxComparatorType is not as simple as it might appear. The nested Box class is currently private and must be exposed for a comparison function because the Box objects are the inputs to the comparison. The Box class also currently has members to support minimum area computations, but minimum width requires other support.

My plan is to factor out the rotating calipers code as a separate class RotatingCalipers with a nested base-class Box object containing only those members needed for identifying the pairs of antipodal vertices/edges. Classes that use RotatingCalipers will be MinimumAreaBox2 and MinimumWidthBox2 with nested derived classes for Box that contain support for the particular feature to be minimized. These derived classes hopefully will be shallow in the sense that adding other such classes for box features is relatively simple.

Hopefully the refactoring will take no longer than about one week of development.

**Comment by davideberly** [comment]
I have posted code for your feature request. Refactoring MinimumAreaBox2 was not the correct approach because it has a pair of rotating calipers that work together. Instead I added the class RotatingCalipers that identifies all the vertex-edge antipodal pairs. The new class MinimalWidthPoints2 uses RotatingCalipers to compute the width of the input point set. It returns a 2D oriented box whose extent[0] is the minimum width. The sample GTE/Samples/Geometrics/MinimumAreaBox2D was modified to display the minimum-area and minimum-width boxes for comparison. The example shows that the two boxes can be different.
